In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn import preprocessing
from sklearn import metrics
from sklearn.linear_model import LinearRegression

Importing, Reading, and Printing Data

In [37]:
df = pd.DataFrame()
df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")
df[0::10]

Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
10    63.095   8117     589      4   30         14        1     1407
20    17.700   2860     235      4   63         16        0       89
30    34.142   5666     413      4   47          5        1      863
40    34.950   3327     253      3   54         14        0       50
50    36.362   5183     376      3   49         15        1      654
60    35.510   5198     364      2   35         20        0      631
70    24.889   3954     318      4   75         12        1      357
80    27.369   3449     288      3   40          9        1      162
90    20.191   5767     431      4   42         16        1     1023
100   21.153   3736     256      1   41         11        0      298
110   34.537   3271     250      3   57         17        1       47
120   27.241   1402     128      2   67         15        1        0
130   23.793   3821     281      4   56         12        1      868
140   39.831   6045     459      3   32         12        1     1425
150   63.931   5728     435      3   28         14        1      581
160   12.238   4865     381      5   67         11        0      836
170   34.772   2021     167      3   57          9        0        0
180   10.635   3584     294      5   69         16        1      423
190   55.187   5352     385      4   50         17        1      538
200   23.949   5343     383      2   40         18        1      829
210   24.543   3206     243      2   62         12        1       95
220   44.847   5765     437      3   53         13        0     1246
230   33.214   5137     387      3   59          9        0      661
240   29.705   3351     262      5   71         14        1      148
250   10.363   2430     191      2   47         18        1        0
260   67.937   5184     383      4   63         12        1      345
270   15.866   3085     217      1   39         13        0      136
280   53.401   5319     377      3   35         12        0      541
290   26.370   3235     268      5   78         11        1      159
300   21.786   4632     355      1   50         17        1      580
310   33.694   4891     369      1   52         16        0     1036
320   16.279   1160     126      3   78         13        1        5
330   20.405   4543     329      2   72         17        0     1054
340   27.794   3807     301      4   35          8        1      320
350   30.002   1561     155      4   70         13        1        0
360   53.566   5891     434      4   82         10        0      712
370   35.610   6135     466      4   40         12        0      992
380  115.123   7760     538      3   83         14        0      661
390  135.118  10578     747      3   81         15        1     1393

Here we label our data in an X and Y array

In [27]:
feature_columns = ['Income','Limit','Rating','Cards','Age','Education','Married']
x = df[feature_columns]
y = df['Balance']

Checking X and Y

In [38]:
print(x.shape)
print(y.shape)

(400, 7)
(400,)


In [40]:
x[0::10]

Income  Limit  Rating  Cards  Age  Education  Married
0     14.891   3606     283      2   34         11        1
10    63.095   8117     589      4   30         14        1
20    17.700   2860     235      4   63         16        0
30    34.142   5666     413      4   47          5        1
40    34.950   3327     253      3   54         14        0
50    36.362   5183     376      3   49         15        1
60    35.510   5198     364      2   35         20        0
70    24.889   3954     318      4   75         12        1
80    27.369   3449     288      3   40          9        1
90    20.191   5767     431      4   42         16        1
100   21.153   3736     256      1   41         11        0
110   34.537   3271     250      3   57         17        1
120   27.241   1402     128      2   67         15        1
130   23.793   3821     281      4   56         12        1
140   39.831   6045     459      3   32         12        1
150   63.931   5728     435      3   28         14        1
160   12.238   4865     381      5   67         11        0
170   34.772   2021     167      3   57          9        0
180   10.635   3584     294      5   69         16        1
190   55.187   5352     385      4   50         17        1
200   23.949   5343     383      2   40         18        1
210   24.543   3206     243      2   62         12        1
220   44.847   5765     437      3   53         13        0
230   33.214   5137     387      3   59          9        0
240   29.705   3351     262      5   71         14        1
250   10.363   2430     191      2   47         18        1
260   67.937   5184     383      4   63         12        1
270   15.866   3085     217      1   39         13        0
280   53.401   5319     377      3   35         12        0
290   26.370   3235     268      5   78         11        1
300   21.786   4632     355      1   50         17        1
310   33.694   4891     369      1   52         16        0
320   16.279   1160     126      3   78         13        1
330   20.405   4543     329      2   72         17        0
340   27.794   3807     301      4   35          8        1
350   30.002   1561     155      4   70         13        1
360   53.566   5891     434      4   82         10        0
370   35.610   6135     466      4   40         12        0
380  115.123   7760     538      3   83         14        0
390  135.118  10578     747      3   81         15        1

In [39]:
y[::10]

0       333
10     1407
20       89
30      863
40       50
50      654
60      631
70      357
80      162
90     1023
100     298
110      47
120       0
130     868
140    1425
150     581
160     836
170       0
180     423
190     538
200     829
210      95
220    1246
230     661
240     148
250       0
260     345
270     136
280     541
290     159
300     580
310    1036
320       5
330    1054
340     320
350       0
360     712
370     992
380     661
390    1393
Name: Balance, dtype: int64

Here we normalize our X and print to check the results

In [23]:
norm_x = preprocessing.scale(x) 
norm_x

array([[-0.86158299, -0.48999879, -0.46553881, ..., -1.2576741 ,
        -0.78492991,  0.79539491],
       [ 1.72743711,  0.82826106,  0.82870309, ...,  1.5284506 ,
         0.49658831,  0.79539491],
       [ 1.68675551,  1.01478681,  1.02931059, ...,  0.88996369,
        -0.78492991, -1.25723711],
       ...,
       [ 0.35946155, -0.24491264, -0.21963285, ...,  0.65778663,
        -0.46455035,  0.79539491],
       [-0.21280808, -0.95891584, -1.05441888, ..., -0.67723146,
        -0.1441708 ,  0.79539491],
       [-0.75334493,  0.34199278,  0.38866085, ...,  0.48365384,
        -2.06644812, -1.25723711]])

Next we split our data into training and testing sets.

In [24]:
x_train,x_test,y_train,y_test = train_test_split(x, y ,test_size = 0.24, random_state = 9)

Next we instantiate an object of the Linear Regression Class and fit the model to the training set.

In [25]:
my_linReg = LinearRegression()
my_linReg.fit(x_train, y_train)

LinearRegression()

Here we print out the Theta

In [28]:
print(my_linReg.intercept_)
print(feature_columns)
print(my_linReg.coef_)

-497.0869076583074
['Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education', 'Married']
[ -7.5197188    0.15640406   1.6355104   18.19349659  -1.00062374
   3.73501837 -20.17447952]


"Cards" had the largest coefficient at 18.19349659, while " Married" had the lowest coefficient at -20.17447952

Next we Calculate the RMSE for the

In [34]:
y_predict = my_linReg.predict(x_test)
mse = metrics.mean_squared_error(y_test, y_predict)
rmse = np.sqrt(mse)
#print(y_predict)
print("RMSE:", rmse)

RMSE: 143.2464973964321


Next we use 10-fold Cross Validation to evaluate our Linear Regression and print out the mean for our rmse list of values.

In [36]:
mseList = cross_val_score(my_linReg, x, y, cv=10, scoring='neg_mean_squared_error')
positiveMseList= -mseList #get positive values
rmseList = np.sqrt(positiveMseList)
print("RMSE List Results:", rmseList)
print("RMSE List Results Mean:", rmseList.mean())

RMSE List Results: [153.7754992  178.89394627 188.31527914 193.20353826 119.75525901
 183.38039983 177.85385343 111.76315739 144.04725656 152.331702  ]
RMSE List Results Mean: 160.33198910744073
